File: computing/interview/design-patterns/structural/proxy/csharp/ProxyDemo.cs
using System;

namespace Interview.DesignPatterns.Structural.Proxy
{
    public interface IImage
    {
        void Display();
    }

    public sealed class RealImage : IImage
    {
        private readonly string _filename;

        public RealImage(string filename)
        {
            _filename = filename;
            Console.WriteLine($"Loading image from disk: {_filename}");
        }

        public void Display()
        {
            Console.WriteLine($"Displaying image: {_filename}");
        }
    }

    public sealed class LazyImageProxy : IImage
    {
        private readonly string _filename;
        private RealImage? _realImage;

        public LazyImageProxy(string filename)
        {
            _filename = filename;
        }

        public void Display()
        {
            _realImage ??= new RealImage(_filename);
            _realImage.Display();
        }
    }

    public static class Demo
    {
        public static void Main()
        {
            IImage proxy = new LazyImageProxy("photo.png");
            Console.WriteLine("First display triggers load:");
            proxy.Display();
            Console.WriteLine("Second display reuses loaded image:");
            proxy.Display();
        }
    }
}

File: computing/interview/design-patterns/structural/proxy/java/ProxyDemo.java
package interview.designpatterns.structural.proxy;

interface Image {
    void display();
}

final class RealImage implements Image {
    private final String filename;

    RealImage(String filename) {
        this.filename = filename;
        System.out.printf("Loading image from disk: %s%n", filename);
    }

    @Override
    public void display() {
        System.out.printf("Displaying image: %s%n", filename);
    }
}

final class LazyImageProxy implements Image {
    private final String filename;
    private RealImage realImage;

    LazyImageProxy(String filename) {
        this.filename = filename;
    }

    @Override
    public void display() {
        if (realImage == null) {
            realImage = new RealImage(filename);
        }
        realImage.display();
    }
}

public final class ProxyDemo {
    private ProxyDemo() {}

    public static void main(String[] args) {
        Image proxy = new LazyImageProxy("photo.png");
        System.out.println("First display triggers load:");
        proxy.display();
        System.out.println("Second display reuses loaded image:");
        proxy.display();
    }
}
